### PR TITLE
drivers/sensor: lis2dh: use gpio_dt_spec oriented macros

### PR DESCRIPTION
**Pull request by avisconti**
Use gpio_dt_spec structure and related macros for both drdy
and AnyMotion interrupts, to have a more compat, readable and
safe code.

This commit also fixes the bug #34794.

### CODE REVIEW

**Comment by avisconti** [comment]
I added DNM as I'm not sure it fixes #34794 (not able to replicate it). It's something that should go in anyway, but I want also to be sure it fixes that issue.

**Comment by avisconti** [comment]
@mbolivar-nordic I would like to get your opinion on device_is_ready() usage when the pointer is NULL.

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic I would like to get your opinion on device_is_ready() usage when the pointer is NULL.

What opinion, sorry? We discussed making device_is_ready() return false when the argument is NULL and I think that makes sense.

**Comment by avisconti** [comment]
> > @mbolivar-nordic I would like to get your opinion on device_is_ready() usage when the pointer is NULL.
> 
> What opinion, sorry? We discussed making device_is_ready() return false when the argument is NULL and I think that makes sense.

So is device_is_ready() currently returning false if arg is NULL? Is it really the case?

**Comment by mbolivar-nordic** [comment]
> So is device_is_ready() currently returning false if arg is NULL? Is it really the case?

Sorry, no, I'm saying it would make sense for that to work. You cannot pass NULL to device_is_ready() right now, at least the last time I checked.

**Comment by avisconti** [comment]
> > So is device_is_ready() currently returning false if arg is NULL? Is it really the case?
> 
> Sorry, no, I'm saying it would make sense for that to work. You cannot pass NULL to device_is_ready() right now, at least the last time I checked.

ok, I'll do in a separate PR

**Comment by pabigot** [comment]
> ok, I'll do in a separate PR

#32367 is somewhat relevant but isn't going anywhere.  So is #32587 but it needs a shepherd.

**Comment by mbolivar-nordic** [comment]
> ok, I'll do in a separate PR

for anyone watching: https://github.com/zephyrproject-rtos/zephyr/pull/35268#pullrequestreview-659096162

**Review by MaureenHelm — COMMENTED** [review]
> > ok, I'll do in a separate PR
> 
> for anyone watching: #35268

This PR depends on that one, right?

**Comment by avisconti** [comment]
> > > ok, I'll do in a separate PR
> > 
> > 
> > for anyone watching: #35268
> 
> This PR depends on that one, right?

@MaureenHelm 
Actually no, there is no strict dependency between this PR and #35268.
But debugging an issue for which this PR is (maybe) a solution (and for sure a code semplification), I also observed yesterday that there is a device_is_ready() misusage (actually quite spread over zephyr code).

File: drivers/sensor/lis2dh/lis2dh.c [hunk 1]
```diff
@@ -417,28 +417,23 @@ int lis2dh_init(const struct device *dev)
 	})
 
 #ifdef CONFIG_LIS2DH_TRIGGER
-#define LIS2DH_HAS_IRQ_IDX(inst, idx)					\
-		DT_INST_PROP_HAS_IDX(inst, irq_gpios, idx)
+#define LIS2DH_SET_TRIG_BY_IDX(id, prop, idx)				\
+	COND_CODE_1(DT_INST_PROP_HAS_IDX(id, prop, idx),		\
+		    (true),						\
+		    (false))
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.c:423]: Yes, of course. There is no need for LIS2DH_SET_TRIG_BY_IDX at all, I can simply use DT_INST_PROP_HAS_IDX. The code was initially different and when I simplified it I came up with this stupid stuff. Yesterday was a bad day for me... sorry.

[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.c]: Thanks for pointing out!

**Comment by mbolivar-nordic** [comment]
>  there is a device_is_ready() misusage (actually quite spread over zephyr code).

@avisconti do you have pointers for this?

I wonder if there is some way to annotate the function so coverity knows it can't take null, and likewise annotate things like DEVICE_DT_GET_ANY so coverity knows they may return null, to start to get some static coverage for this.

**Comment by avisconti** [comment]
> > there is a device_is_ready() misusage (actually quite spread over zephyr code).
> 
> @avisconti do you have pointers for this?
> 
> I wonder if there is some way to annotate the function so coverity knows it can't take null, and likewise annotate things like DEVICE_DT_GET_ANY so coverity knows they may return null, to start to get some static coverage for this.

I'm answering on #35268. This PR is somehow related, but its intent is more a code simplification.

**Comment by galak** [comment]
@avisconti / @mbolivar-nordic did you guys come to resolution on this?

**Comment by mbolivar-nordic** [comment]
> @avisconti / @mbolivar-nordic did you guys come to resolution on this?

I didn't look at this PR since I thought it was a new feature; I missed it was a bugfix also. Looking at it now

File: drivers/sensor/lis2dh/lis2dh.h [hunk 2]
```diff
@@ -193,12 +193,10 @@ struct lis2dh_config {
...
-	const char *irq1_dev_name;
-	gpio_pin_t irq1_pin;
-	gpio_dt_flags_t irq1_flags;
-	const char *irq2_dev_name;
-	gpio_pin_t irq2_pin;
-	gpio_dt_flags_t irq2_flags;
+	const struct gpio_dt_spec gpio_drdy;
+	const struct gpio_dt_spec gpio_int;
+	bool trig_drdy_en;
+	bool trig_int_en;
```
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dh/lis2dh.h:199]: I think these fields are not needed and can be removed.

It seems like `cfg->trig_drdy_en` is the same thing as `cfg->gpio_drdy.port != NULL`, and similarly for `trig_int_en`.

File: drivers/sensor/lis2dh/lis2dh_trigger.c [hunk 3]
```diff
@@ -385,11 +381,9 @@ int lis2dh_init_interrupt(const struct device *dev)
 	 */
 
 	/* setup data ready gpio interrupt */
-	lis2dh->gpio_int1 = device_get_binding(cfg->irq1_dev_name);
-	if (lis2dh->gpio_int1 == NULL) {
-		LOG_INF("Cannot get pointer to irq1_dev_name");
-		status = 0;
-		goto end;
+	if (!device_is_ready(cfg->gpio_drdy.port)) {
```
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:384]: Is this right? It seems like the following could happen:

- lis2dh_init(dev) is called
- CONFIG_LIS2DH_TRIGGER=y, so lis2dh_init_interrupt(dev) is called
- dev has no drdy GPIO set up in devicetree, so `cfg->gpio_drdy.port` is NULL
- device_is_ready() returns 0 for NULL argument, so lis2dh_init_interrupt() returns -EINVAL
- lis2dh_init(dev) fails

I'm wondering if the drdy GPIO is not defined for just one instance, even though CONFIG_LIS2DH_TRIGGER=y globally, we should fail the entire device init or not.

File: drivers/sensor/lis2dh/lis2dh_trigger.c [hunk 4]
```diff
@@ -407,56 +401,54 @@ int lis2dh_init_interrupt(const struct device *dev)
 #endif
 
 	/* data ready int1 gpio configuration */
-	status = gpio_pin_configure(lis2dh->gpio_int1, cfg->irq1_pin,
-				    GPIO_INPUT | cfg->irq1_flags);
+	status = gpio_pin_configure_dt(&cfg->gpio_drdy, GPIO_INPUT);
 	if (status < 0) {
-		LOG_ERR("Could not configure gpio %d", cfg->irq1_pin);
+		LOG_ERR("Could not configure gpio %d", cfg->gpio_drdy.pin);
```
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:406]: Error logging is dropping the device name here, which will make debugging harder if two ports have the same pin number and are used for DRDY. Not caused by this PR though. Note that the LOG_INF() at line 420 gets this right.

File: drivers/sensor/lis2dh/lis2dh_trigger.c [hunk 5]
```diff
@@ -407,56 +401,54 @@ int lis2dh_init_interrupt(const struct device *dev)
...
 
 	/*
 	 * Setup INT2 (for Any Motion) if defined in DT
 	 */
+
 	/* setup any motion gpio interrupt */
-	lis2dh->gpio_int2 = device_get_binding(cfg->irq2_dev_name);
-	if (lis2dh->gpio_int2 == NULL) {
-		LOG_INF("Cannot get pointer to irq2_dev_name");
+	if (!device_is_ready(cfg->gpio_int.port)) {
```
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:427]: Is it correct for NULL ports?

File: drivers/sensor/lis2dh/lis2dh_trigger.c [hunk 6]
```diff
@@ -407,56 +401,54 @@ int lis2dh_init_interrupt(const struct device *dev)
...
 		goto end;
 	}
 
 	/* any motion int2 gpio configuration */
-	status = gpio_pin_configure(lis2dh->gpio_int2, cfg->irq2_pin,
-				    GPIO_INPUT | cfg->irq2_flags);
+	status = gpio_pin_configure_dt(&cfg->gpio_int, GPIO_INPUT);
 	if (status < 0) {
-		LOG_ERR("Could not configure gpio %d", cfg->irq2_pin);
+		LOG_ERR("Could not configure gpio %d", cfg->gpio_int.pin);
```
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:436]: This also drops the port name.

File: drivers/sensor/lis2dh/lis2dh.h (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.h:199]: Yes, I could test (cfg->gpio_drdy.port != NULL). It is the same thing in fact.
OK, I'll remove the flags.

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 3 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: So, I would go in this way if you agree...

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 6 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:436]: ack

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 5 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:427]: It is returning false, but not giving any error (status = 0)

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 3 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: Actually I think that testing gpio_drdy.port would be enough. That's because I'm not sure it is possible to configure only gpio_int in DT. I mean, it would be perfectly legit from sensor point of view, but from DT standpoint IDX=1 is valid only if IDX=0 is present. Isn't it?

I tried:
'''
irq-gpios = ,  <&arduino_header 4 GPIO_ACTIVE_HIGH>;
'''

:D

Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: >  from DT standpoint IDX=1 is valid only if IDX=0 is present. Isn't it?

The secret sauce is `irq-gpios = <0>, <&arduino_header 4 GPIO_ACTIVE_HIGH>;`

Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: > To be honest I think that a flag would be more clear, but I understand also your point to be smaller in size.

Could use a static inline `is_trig_enabled` or so for readability.

Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: > The secret sauce is 

I learned this trick here: https://elixir.bootlin.com/linux/latest/source/include/linux/of_gpio.h#L92

We started supporting it in edtlib in 38ede5aded05964054e1706058159bef94bf2d7d though.

Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: In this case I should test both.

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 5 above)
Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:428]: > It is returning false, but not giving any error (status = 0)

Then why LOG_ERR?

Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:428]: > LOG_DBG

Sounds good

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 3 above)
Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: Test both what, sorry? You can test each index independently like you're doing now if I understand your purpose correctly.

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 5 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:432]: ok, then. Thanks for your patience.

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 3 above)
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: My last code is like:
```
if (cfg->gpio_drdy.port != NULL || cfg->gpio_int.port != NULL) {
    status = lis2dh_init_interrupt(dev);
    if (status < 0) {
        LOG_ERR("Failed to initialize interrupts.");
        return status;
    }
}
```

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 5 above)
Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:432]: Thanks for yours!

File: drivers/sensor/lis2dh/lis2dh_trigger.c (hunk 3 above)
Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: OK, I see what you mean by "test both": since lis2dh_init_interrupt() is responsible for initializing both interrupts, so you need to test if either of these pointers is not NULL.

But I'm not sure I follow the logic in that function anymore.

I think there's a difference between these 2 things:

1. `(cfg->gpio_drdy.port != NULL) && (device_is_ready(cfg->gpio_drdy.port) == false))`
2. `cfg->gpio_drdy.port == NULL`

I think that 1. should be an error but 2. is not an error.

But 1. can still happen inside lis2dh_init_interrupt if `cfg->gpio_drdy.port != NULL && cfg->gpio_int.port != NULL`, right?

And lis2dh_interrupt still returns zero, which seems wrong, since 1. should be an error.

Am I missing something?

Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: So, both means gpio_drdy or gpio_int is configured.

Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: > OK, I see what you mean by "test both": since lis2dh_init_interrupt() is responsible for initializing both interrupts, you need to test if either of these pointers is not NULL.
> 
> But I'm not sure I follow the logic in that function anymore.
> 
> I think there's a difference between these 2 things:
> 
> 1. `(cfg->gpio_drdy.port != NULL) && (device_is_ready(cfg->gpio_drdy.port) == false))`
> 2. `cfg->gpio_drdy.port == NULL`
> 
> I think that 1. should be an error but 2. is not an error.
> 
> But 1. can still happen inside lis2dh_init_interrupt if `cfg->gpio_drdy.port != NULL && cfg->gpio_int.port != NULL`, right?
> 
> And lis2dh_interrupt still returns zero, which seems wrong, since 1. should be an error.
> 
> Am I missing something?

What you are saying is correct. It may happen that device is not ready even if gpio_drdy.port is valid. In this case code is not returning an error even if it should. I have similar issue in all sensor drivers. I think that I need an extra test if device_is_ready return false, as the API was changed in #35268. Actually everybody using this API should...

Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: I'll resubmit the PR

Comment (mbolivar-nordic) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:384]: > Actually everybody using this API should...

Well, before, nobody should have been passing NULL to device_is_ready, so existing code which did before the API changed was just wrong, so I guess it's at least not a new issue if it's still wrong.

>  I'll resubmit the PR

Sounds good, I've unresolved this comment in the meantime.

File: drivers/sensor/lis2dh/lis2dh_trigger.c [hunk 7]
```diff
@@ -385,11 +381,16 @@ int lis2dh_init_interrupt(const struct device *dev)
 	 */
 
 	/* setup data ready gpio interrupt */
-	lis2dh->gpio_int1 = device_get_binding(cfg->irq1_dev_name);
-	if (lis2dh->gpio_int1 == NULL) {
-		LOG_INF("Cannot get pointer to irq1_dev_name");
+	if (!device_is_ready(cfg->gpio_drdy.port)) {
+		/* API may return false even when ptr is NULL */
+		if (cfg->gpio_drdy.port != NULL) {
+			LOG_ERR("Cannot get pointer to gpio_drdy device");
```
Comment (mbolivar-nordic) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:387]: One last thing, I guess in this case you do have a pointer, it's just not ready. Usually we do something like this then:

```
LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
```

Same for the other device.

Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:387]: sounds good

**Review by mbolivar-nordic — APPROVED** [review]

File: drivers/sensor/lis2dh/lis2dh.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.c]: > This could be simplified to just `DT_INST_PROP_HAS_IDX(id, prop, idx)`. I don't think you need the `COND_CODE_1`

In latest submitted code I took care also for this suggestions. Thanks!

**Comment by avisconti** [comment]
@MaureenHelm 
Hello Maureen, can you take a look to this PR again?

**Review by MaureenHelm — APPROVED** [review]
